var/spack/repos/builtin/packages/go/package.py

```diff
@@ -104,14 +104,7 @@ def install(self, spec, prefix):
                 shutil.copy2(f, os.path.join(prefix, f))
 
     def setup_environment(self, spack_env, run_env):
-        try:
-            spack_env.set('GOROOT_FINAL', self.spec.prefix)
-            spack_env.set('GOROOT_BOOTSTRAP', self.spec['go-bootstrap'].prefix)
-        except KeyError:
-            # Needed because we try to get a build-only
-            # dependency, and this may fail during module 
-            # file creation
-            pass
+        spack_env.set('GOROOT_FINAL', self.spec.prefix)
```

Comment:
Using those two variables only at build time is exactly what is done in this PR. 

There are two callbacks that can be overridden: `setup_dependent_environment` and `setup_environment`. The former lets a package inject variables in the environment of a package that *depends on* you, the latter lets you set your own environment. Both the callbacks take two arguments: `spack_env` is the build-time environment, `run_env` is the run-time environment (a.k.a. what gets written in module files).